lib/galaxy/tools/loader_directory.py

```diff
@@ -264,6 +265,8 @@ def _find_files(directory, pattern='*'):
 
     matches = []
     for root, dirnames, filenames in os.walk(directory):
+        # exclude some directories (like .hg) from traversing
+        dirnames[:] = [dir for dir in dirnames if dir not in EXCLUDE_WALK_DIRS]
```

Comment:
You probably don't need the `[:]`

Reply:
Mh, I thought we need it according to https://docs.python.org/2/library/os.html#os.walk:

```
When topdown is True, the caller can modify the dirnames list in-place (perhaps using del or slice assignment), and walk() will only recurse into the subdirectories whose names remain in dirnames; this can be used to prune the search, impose a specific order of visiting, or even to inform walk() about directories the caller creates or renames before it resumes walk() again.
```

Reply:
The `[:]` makes sense on the left side only if you add some numbers around the colon (e.g. `l[3:5]`) to modify a part of the list, here `dirnames` is completely replaced.

Reply:
@nsoranzo I don't think you are right. One operation is pointing a variable at a new list and the other operation is replacing the contents of a list. Check out the following example:

```
>>> x = [1, 2, 3]
>>> y = x
>>> y = [4, 5, 6]
>>> x
[1, 2, 3]
>>> y = x
>>> y[:] = [4, 5, 6]
>>> x
[4, 5, 6]
```

If you just replace ``dirnames`` to point at a new list, there is no way ``os.walk`` internals would know about this. If you change the list itself though - it can use the new contents.

Reply:
Mmmh, I'll blame it on Monday morning, you're all right, sorry for the noise.

Reply:
In the second case,
```
>>> y = x
>>> y[:] = [4, 5, 6]
>>> x
[4, 5, 6]
```
no new object is created. `x `and `y `point to the same object and that object we mutate using `y[:]`. But in the first case, we just have another pointer (`y`) to the object already pointed to by `x` (`y = x`). In line `y = [4, 5, 6]`
`y` points to another object. I think calling a method like an append or slicing, mutates a list.